### PR TITLE
os/bluestore: optimize mutex contention

### PR DESCRIPTION
**Pull request by majianpeng**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by majianpeng** [comment]
@markhpc . can you review this pr?

**Comment by gregsfortytwo** [comment]
The Throttle change looks fine to me.

**Comment by majianpeng** [comment]
@markhpc ,  i did 4k rw performance test with one nvme ssd. but the performance doesn't improve.

**Comment by majianpeng** [comment]
@ifed01 . ping.

**Comment by markhpc** [comment]
I'm afraid I probably won't have time to look at this as I'm pretty booked up for the next couple of weeks.  Maybe @ifed01 can?

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -11166,59 +11166,85 @@ bluestore_deferred_op_t *BlueStore::_get_deferred_op(
...
-      !txc->osr->deferred_running) {
-    _deferred_submit_unlock(txc->osr.get());
-  } else {
-    deferred_lock.unlock();
+
+  {
+    txc->osr->deferred_lock.lock();
+    ++deferred_queue_size;
+    txc->osr->deferred_pending = tmp;
+    if (!txc->osr->in_deferred_queue) {
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:11196]: can we simplify this by using local flag showing whether deferred_pending is newly allocated container (i.e. deferred_pending was nullptr on the entry to the func)? Hence no need for a new member in OSD...
hence the check to be something like
if (!was_deferred_pending && !deferred_running) {
...
}

**Review by ifed01 — COMMENTED** [review]
Sorry for that long review. This might make sense indeed. The effect on deferred ops would be rather less visible since 4k doesn't trigger deferred ops any more

**Comment by majianpeng** [comment]
@ifed01 . update by your suggestion. please review.  Thanks much!

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -12131,59 +12131,84 @@ bluestore_deferred_op_t *BlueStore::_get_deferred_op(
...
-      !txc->osr->deferred_running) {
-    _deferred_submit_unlock(txc->osr.get());
-  } else {
-    deferred_lock.unlock();
+
+  {
+    txc->osr->deferred_lock.lock();
+    ++deferred_queue_size;
+    txc->osr->deferred_pending = tmp;
+    if (!txc->osr->deferred_running && (tmp->txcs.size() == 1)) {
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:12162]: nit: may be leave a comment about (tmp->txc.size() == 1) meaning...

**Review by ifed01 — APPROVED** [review]

**Comment by majianpeng** [comment]
retest this please

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-12-30_06:47:01-rados-wip-kefu-testing-2020-12-30-1123-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-12-31_04:10:08-rados-wip-kefu-testing-2020-12-30-1123-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-12-31_09:50:24-rados-wip-kefu-testing-2020-12-31-1427-distro-basic-smithi/

test failures tracked by

- https://tracker.ceph.com/issues/46323

**Pull request merged**